### PR TITLE
Add an implementation for the passenger API

### PR DESCRIPTION
**Pull request by syldium**
# Description
This PR resolves the TODOs left in #519 by implementing the passenger-related API.

An entity can now have a vehicle and have multiple passengers. This fires four new events:

- Bukkit's `VehicleEnterEvent` only when the vehicle implements `Vehicle`
- Spigot's `EntityMountEvent`
- Bukkit's `VehicleExitEvent` only when the vehicle implements `Vehicle`
- Spigot's `EntityDismountEvent`

In Vanilla/CraftBukkit, the rider always sets the reference to their vehicle first, then the vehicle updates its passenger list:
```mermaid
flowchart TB
    subgraph Passenger
    set[set vehicle]
    unset[unset vehicle]
    end
    subgraph Vehicle
    assertions
    events
    add[add passenger]
    end
    set --> assertions
    assertions --> events
    assertions -->|fail| unset
    events -->|cancelled| unset
    events --> add
```

Because of this, this sample code would dismount the sheep even if it's not the vehicle:
```java
Sheep a = world.spawn(location, Sheep.class);
Sheep b = world.spawn(location, Sheep.class);
a.addPassenger(b);
player.removePassenger(b); // The player is not the vehicle of the sheep, but can still have it dismounted from the real vehicle.
```
This behavior is replicated in this implementation.

Since #519, PaperMC/Paper@5deafd1969b3c521af9cabb46822ecb01e6ce498 has been added to Paper, allowing greater control over teleports with passengers. Teleportation methods have been updated to accept these new parameters. Note that these methods have been annotated with [`@ApiStatus.Experimental`](https://github.com/JetBrains/java-annotations/blob/d438b1132495b7fc5c272289441781d9a0275e4c/common/src/main/java/org/jetbrains/annotations/ApiStatus.java#L54).

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Unit tests added.
- [x] Code follows existing code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, update the patch version. (if the current version is `0.4.1`, the new version should be `0.4.2`)
- If the PR adds a new feature to MockBukkit, update minor version. (if the current version is `0.4.1`, the new version should be `0.5.0`)

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java [hunk 1]
```diff
@@ -785,22 +852,69 @@ public void playEffect(@NotNull EntityEffect type)
...
+			}
+		}
+		if (!new EntityMountEvent(entity, this).callEvent())
+		{
+			return false;
+		}
+		return this.passengers.add(entity);
+	}
+
+	private boolean tryRemovingPassenger(@NotNull Entity entity)
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:926]: This should have a Javadoc

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java [hunk 2]
```diff
@@ -785,22 +852,69 @@ public void playEffect(@NotNull EntityEffect type)
...
 	@Override
-	public Entity getVehicle()
+	public @Nullable Entity getVehicle()
 	{
-		// TODO Auto-generated constructor stub
-		throw new UnimplementedOperationException();
+		return this.vehicle;
+	}
+
+	private boolean tryAddingPassenger(@NotNull Entity entity)
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:894]: This should have a Javadoc

File: src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java [hunk 3]
```diff
@@ -566,13 +566,13 @@ public boolean getCanPickupItems()
 	}
 
 	@Override
-	public boolean teleport(@NotNull Location location, PlayerTeleportEvent.@NotNull TeleportCause cause)
+	public boolean teleport(@NotNull Location location, PlayerTeleportEvent.@NotNull TeleportCause cause, boolean ignorePassengers, boolean dismount)
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java]: Why was this changed?

**Review by Insprill — CHANGES_REQUESTED** [review]
Looks good, just a couple small things

File: src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java (hunk 3 above)
Comment (syldium) [reply src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java]: In order to maintain the health check when using the new Paper's API:
```java
// Redirects to teleport(Location, TeleportCause) that checks health, then uses teleport(Location, TeleportCause, boolean, boolean).
livingEntity.teleport(loc);
// Would skip the health check since this method is only implemented in the super class.
livingEntity.teleport(loc, TeleportCause.PLUGIN, true, true);
```
In fact, the `LivingEntity` class was not modified by PaperMC/Paper@5deafd1969b3c521af9cabb46822ecb01e6ce498, something that has been forgotten?

Comment (Insprill) [reply src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java]: Since it wasn't modified in Paper I feel as if it should not be modified here either to maintain behavior parody. If Paper ever fixes it then we can implement it here.

**Review by Insprill — APPROVED** [review]
LGTM
